File: foreign/typescript/test/support/test-iggy.ts
import { execFile, spawn, type ChildProcess } from "node:child_process"
import { open, mkdtemp, readFile, rm, type FileHandle } from "node:fs/promises"
import { createConnection, createServer } from "node:net"
import { tmpdir } from "node:os"
import { join, resolve } from "node:path"
import { promisify } from "node:util"

const execute = promisify(execFile)
const REPOSITORY_ROOT = resolve(process.cwd(), "../..")
const RESOLVER = join(REPOSITORY_ROOT, "scripts", "resolve-test-iggy-server.sh")

function freePort(): Promise<number> {
  return new Promise((resolvePort, reject) => {
    const server = createServer()
    server.once("error", reject)
    server.listen(0, "127.0.0.1", () => {
      const address = server.address()
      if (address === null || typeof address === "string") {
        server.close()
        reject(new Error("failed to reserve an Iggy test port"))
        return
      }
      server.close((error) => {
        if (error !== undefined) reject(error)
        else resolvePort(address.port)
      })
    })
  })
}

function waitForExit(child: ChildProcess): Promise<void> {
  if (child.exitCode !== null) return Promise.resolve()
  return new Promise((resolveExit) => {
    child.once("exit", () => {
      resolveExit()
    })
  })
}

async function stop(child: ChildProcess | undefined): Promise<void> {
  if (child?.exitCode !== null) return
  child.kill("SIGTERM")
  let timer: NodeJS.Timeout | undefined
  const timeout = new Promise<void>((resolveTimeout) => {
    timer = setTimeout(() => {
      if (child.exitCode === null) child.kill("SIGKILL")
      resolveTimeout()
    }, 5_000)
  })
  try {
    await Promise.race([waitForExit(child), timeout])
  } finally {
    if (timer !== undefined) clearTimeout(timer)
  }
}

async function resolveBinary(): Promise<string> {
  const configured = process.env["LASER_TEST_IGGY_SERVER"]
  if (configured !== undefined && configured.length > 0) return configured
  const { stdout } = await execute(RESOLVER, [], {
    cwd: REPOSITORY_ROOT,
    env: process.env,
    timeout: 310_000
  })
  const binary = stdout.trim()
  if (binary.length === 0) throw new Error("Iggy test server resolver returned no path")
  return binary
}

export class TestIggy {
  private child: ChildProcess | undefined
  private closed = false

  private constructor(
    private readonly binary: string,
    private readonly directory: string,
    private readonly port: number,
    private readonly log: FileHandle
  ) {}

  get endpoint(): string {
    return `iggy:iggy@127.0.0.1:${String(this.port)}`
  }

  static async start(): Promise<TestIggy> {
    const binary = await resolveBinary()
    const directory = await mkdtemp(join(tmpdir(), "laser-iggy-"))
    const port = await freePort()
    const log = await open(join(directory, "test-server.log"), "w")
    const server = new TestIggy(binary, directory, port, log)
    try {
      await server.spawn()
      return server
    } catch (error) {
      await server.close()
      throw error
    }
  }

  async restart(): Promise<void> {
    await stop(this.child)
    this.child = undefined
    await this.spawn()
  }

  async close(): Promise<void> {
    if (this.closed) return
    this.closed = true
    await stop(this.child)
    await this.log.close()
    await rm(this.directory, { recursive: true, force: true })
  }

  private async spawn(): Promise<void> {
    const logPath = join(this.directory, "test-server.log")
    this.child = spawn(this.binary, [], {
      env: {
        ...process.env,
        IGGY_SYSTEM_PATH: this.directory,
        IGGY_TCP_ADDRESS: `127.0.0.1:${String(this.port)}`,
        IGGY_HTTP_ENABLED: "false",
        IGGY_QUIC_ENABLED: "false",
        IGGY_WEBSOCKET_ENABLED: "false",
        IGGY_ROOT_USERNAME: "iggy",
        IGGY_ROOT_PASSWORD: "iggy",
        IGGY_SHARD_RUNTIME_CAPACITY: "256",
        IGGY_SYSTEM_SHARDING_RECONCILE_PERIODIC_INTERVAL: "200 ms"
      },
      stdio: ["ignore", this.log.fd, this.log.fd]
    })
    const deadline = Date.now() + 30_000
    while (Date.now() < deadline) {
      if (this.child.exitCode !== null) {
        const contents = await readFile(logPath, "utf8")
        throw new Error(`Iggy test server exited with ${String(this.child.exitCode)}:\n${contents}`)
      }
      const ready = await new Promise<boolean>((resolveReady) => {
        const socket = createConnection({ host: "127.0.0.1", port: this.port })
        socket.once("connect", () => {
          socket.destroy()
          resolveReady(true)
        })
        socket.once("error", () => {
          socket.destroy()
          resolveReady(false)
        })
      })
      if (ready) return
      await new Promise((resolveDelay) => setTimeout(resolveDelay, 100))
    }
    throw new Error("Iggy test server did not accept TCP connections within 30s")
  }
}

export class TestIggyCluster {
  private readonly children: (ChildProcess | undefined)[] = [undefined, undefined, undefined]
  private readonly directories: string[] = []
  private readonly logs: FileHandle[] = []
  private proxy: ReturnType<typeof createServer> | undefined
  private proxyTarget: number

  private constructor(
    private readonly binary: string,
    private readonly tcpPorts: readonly number[],
    private readonly replicaPorts: readonly number[],
    private readonly httpPorts: readonly number[],
    private readonly proxyPort: number
  ) {
    this.proxyTarget = required(tcpPorts[0], "cluster TCP port 0")
  }

  get endpoint(): string {
    return `iggy:iggy@127.0.0.1:${String(this.proxyPort)}`
  }

  static async start(): Promise<TestIggyCluster> {
    const binary = await resolveBinary()
    const ports = await uniquePorts(10)
    const cluster = new TestIggyCluster(
      binary,
      ports.slice(0, 3),
      ports.slice(3, 6),
      ports.slice(6, 9),
      required(ports[9], "cluster proxy port")
    )
    try {
      await cluster.startProxy()
      for (let replicaId = 0; replicaId < 3; replicaId += 1) {
        const directory = await mkdtemp(join(tmpdir(), "laser-iggy-cluster-"))
        cluster.directories.push(directory)
        cluster.logs.push(await open(join(directory, "test-server.log"), "w"))
        cluster.spawn(replicaId)
      }
      await cluster.waitForMesh()
      return cluster
    } catch (error) {
      await cluster.close()
      throw error
    }
  }

  async restartNode(replicaId: number): Promise<void> {
    await this.stopNode(replicaId)
    await this.startNode(replicaId)
  }

  async stopNode(replicaId: number): Promise<void> {
    await stop(this.children[replicaId])
    this.children[replicaId] = undefined
    await this.logs[replicaId]?.close()
    this.logs[replicaId] = await open(
      join(required(this.directories[replicaId], "cluster data directory"), "test-server.log"),
      "w"
    )
  }

  async startNode(replicaId: number): Promise<void> {
    this.spawn(replicaId)
    await this.waitForMesh(replicaId)
  }

  routeEndpointTo(replicaId: number): void {
    this.proxyTarget = required(this.tcpPorts[replicaId], "cluster route target")
  }

  async leaderAndFollower(): Promise<readonly [number, number]> {
    const baseUrl = `http://127.0.0.1:${String(required(this.httpPorts[0], "HTTP port 0"))}`
    const deadline = Date.now() + 30_000
    while (Date.now() < deadline) {
      try {
        const login = await fetch(`${baseUrl}/users/login`, {
          method: "POST",
          headers: { "content-type": "application/json" },
          body: JSON.stringify({ username: "iggy", password: "iggy" })
        })
        if (!login.ok) throw new Error(`login returned ${String(login.status)}`)
        const identity = (await login.json()) as {
          access_token?: { token?: string }
        }
        const token = identity.access_token?.token
        if (token === undefined) throw new Error("login returned no access token")
        const response = await fetch(`${baseUrl}/cluster/metadata`, {
          headers: { authorization: `Bearer ${token}` }
        })
        if (!response.ok) throw new Error(`metadata returned ${String(response.status)}`)
        const metadata = (await response.json()) as {
          nodes?: readonly { name: string; role: string }[]
        }
        const leaderNode = metadata.nodes?.find((node) => node.role === "leader")
        if (leaderNode === undefined) throw new Error("metadata returned no leader")
        const leader = Number.parseInt(leaderNode.name.replace(/^node-/, ""), 10)
        if (!Number.isInteger(leader) || leader < 0 || leader >= 3) {
          throw new Error(`invalid leader name ${leaderNode.name}`)
        }
        const follower = [0, 1, 2].find((node) => node !== leader)
        return [leader, required(follower, "follower node")]
      } catch {
        await new Promise((resolveDelay) => setTimeout(resolveDelay, 100))
      }
    }
    throw new Error("Iggy cluster leader was not discoverable within 30s")
  }

  async close(): Promise<void> {
    for (const child of this.children) await stop(child)
    const proxy = this.proxy
    if (proxy !== undefined) {
      await new Promise<void>((resolveClose) => {
        proxy.close(() => {
          resolveClose()
        })
      })
      this.proxy = undefined
    }
    for (const log of this.logs) await log.close().catch(() => undefined)
    for (const directory of this.directories) {
      await rm(directory, { recursive: true, force: true })
    }
  }

  private startProxy(): Promise<void> {
    return new Promise((resolveListen, reject) => {
      const proxy = createServer((incoming) => {
        const upstream = createConnection({ host: "127.0.0.1", port: this.proxyTarget })
        incoming.pipe(upstream)
        upstream.pipe(incoming)
        upstream.once("error", () => incoming.destroy())
        incoming.once("error", () => upstream.destroy())
      })
      proxy.once("error", reject)
      proxy.listen(this.proxyPort, "127.0.0.1", () => {
        proxy.off("error", reject)
        this.proxy = proxy
        resolveListen()
      })
    })
  }

  private spawn(replicaId: number): void {
    const env: NodeJS.ProcessEnv = {
      ...process.env,
      IGGY_SYSTEM_PATH: this.directories[replicaId],
      IGGY_CLUSTER_ENABLED: "true",
      IGGY_CLUSTER_NAME: "laser-sdk-rolling-restart",
      IGGY_MESSAGE_BUS_RECONNECT_PERIOD: "100ms",
      IGGY_HTTP_ENABLED: "true",
      IGGY_HTTP_ADDRESS: `127.0.0.1:${String(required(this.httpPorts[replicaId], "HTTP port"))}`,
      IGGY_QUIC_ENABLED: "false",
      IGGY_WEBSOCKET_ENABLED: "false",
      IGGY_ROOT_USERNAME: "iggy",
      IGGY_ROOT_PASSWORD: "iggy",
      IGGY_SHARD_RUNTIME_CAPACITY: "256",
      IGGY_SYSTEM_SHARDING_CPU_ALLOCATION: "0..1",
      IGGY_SYSTEM_SHARDING_RECONCILE_PERIODIC_INTERVAL: "200 ms"
    }
    for (let node = 0; node < 3; node += 1) {
      const prefix = `IGGY_CLUSTER_NODES_${String(node)}`
      env[`${prefix}_NAME`] = `node-${String(node)}`
      env[`${prefix}_IP`] = "127.0.0.1"
      env[`${prefix}_REPLICA_ID`] = String(node)
      env[`${prefix}_PORTS_TCP`] = String(this.tcpPorts[node])
      env[`${prefix}_PORTS_TCP_REPLICA`] = String(this.replicaPorts[node])
      env[`${prefix}_PORTS_HTTP`] = String(this.httpPorts[node])
    }
    const log = this.logs[replicaId]
    if (log === undefined)
      throw new Error(`missing log handle for cluster node ${String(replicaId)}`)
    this.children[replicaId] = spawn(this.binary, ["--replica-id", String(replicaId)], {
      env,
      stdio: ["ignore", log.fd, log.fd]
    })
  }

  private async waitForMesh(replicaId?: number): Promise<void> {
    const nodes = replicaId === undefined ? [0, 1, 2] : [replicaId]
    const deadline = Date.now() + 30_000
    while (Date.now() < deadline) {
      let ready = true
      for (const node of nodes) {
        const child = this.children[node]
        const directory = required(this.directories[node], "cluster data directory")
        if (child?.exitCode !== null) {
          const contents = await readFile(join(directory, "test-server.log"), "utf8")
          throw new Error(`Iggy cluster node exited with ${String(child?.exitCode)}:\n${contents}`)
        }
        const contents = await readFile(join(directory, "test-server.log"), "utf8")
        ready = ready && contents.includes("replica mesh complete")
      }
      if (ready) return
      await new Promise((resolveDelay) => setTimeout(resolveDelay, 100))
    }
    throw new Error("Iggy cluster did not form its replica mesh within 30s")
  }
}

function required<T>(value: T | undefined, name: string): T {
  if (value === undefined) throw new Error(`missing ${name}`)
  return value
}

async function uniquePorts(count: number): Promise<number[]> {
  const ports = new Set<number>()
  while (ports.size < count) ports.add(await freePort())
  return [...ports]
}

File: sdk/tests/support/test_iggy.rs
use bytes::Bytes;
use iggy::prelude::*;
use laser_sdk::prelude::Laser;
use std::fs::File;
use std::path::{Path, PathBuf};
use std::process::{Child, Command, Stdio};
use std::sync::Arc;
use std::sync::Mutex;
use std::sync::atomic::{AtomicU16, Ordering};
use std::time::Duration;
use tempfile::TempDir;

const STARTUP_TIMEOUT: Duration = Duration::from_secs(30);
const IGGY_SERVER_ENV: &str = "LASER_TEST_IGGY_SERVER";
const CLUSTER_SIZE: usize = 3;

struct ClusterNodeProcess {
    child: Child,
    data_dir: TempDir,
    log_path: PathBuf,
}

pub struct TestIggyCluster {
    binary: PathBuf,
    nodes: Vec<ClusterNodeProcess>,
    tcp_ports: [u16; CLUSTER_SIZE],
    replica_ports: [u16; CLUSTER_SIZE],
    proxy_port: u16,
    proxy_target: Arc<AtomicU16>,
    proxy_task: tokio::task::JoinHandle<()>,
}

impl TestIggyCluster {
    pub async fn start() -> Self {
        let binary = resolve_server_binary();
        let ports = free_host_ports(CLUSTER_SIZE * 2 + 1);
        let tcp_ports: [u16; CLUSTER_SIZE] = ports[..CLUSTER_SIZE].try_into().expect("TCP ports");
        let replica_ports: [u16; CLUSTER_SIZE] = ports[CLUSTER_SIZE..CLUSTER_SIZE * 2]
            .try_into()
            .expect("replica ports");
        let proxy_port = ports[CLUSTER_SIZE * 2];
        let proxy_target = Arc::new(AtomicU16::new(tcp_ports[0]));
        let proxy_task = spawn_stable_proxy(proxy_port, proxy_target.clone()).await;
        let mut nodes = Vec::with_capacity(CLUSTER_SIZE);
        for replica_id in 0..CLUSTER_SIZE {
            let data_dir = tempfile::tempdir().expect("cluster node data directory");
            let log_path = data_dir.path().join("test-server.log");
            let child = spawn_cluster_node(
                &binary,
                data_dir.path(),
                &log_path,
                replica_id,
                &tcp_ports,
                &replica_ports,
            );
            nodes.push(ClusterNodeProcess {
                child,
                data_dir,
                log_path,
            });
        }
        let cluster = Self {
            binary,
            nodes,
            tcp_ports,
            replica_ports,
            proxy_port,
            proxy_target,
            proxy_task,
        };
        cluster.wait_for_mesh(None).await;
        cluster
    }

    pub fn endpoint(&self) -> String {
        format!("127.0.0.1:{}", self.proxy_port)
    }

    pub fn node_endpoint(&self, replica_id: usize) -> String {
        format!("127.0.0.1:{}", self.tcp_ports[replica_id])
    }

    pub fn route_endpoint_to(&self, replica_id: usize) {
        self.proxy_target
            .store(self.tcp_ports[replica_id], Ordering::Release);
    }

    pub fn stop_node(&mut self, replica_id: usize) {
        let node = self.nodes.get_mut(replica_id).expect("cluster node");
        graceful_stop(&mut node.child);
    }

    pub async fn start_node(&mut self, replica_id: usize) {
        let node = self.nodes.get_mut(replica_id).expect("cluster node");
        node.child = spawn_cluster_node(
            &self.binary,
            node.data_dir.path(),
            &node.log_path,
            replica_id,
            &self.tcp_ports,
            &self.replica_ports,
        );
        self.wait_for_mesh(Some(replica_id)).await;
    }

    pub async fn restart_node(&mut self, replica_id: usize) {
        self.stop_node(replica_id);
        self.start_node(replica_id).await;
    }

    async fn wait_for_mesh(&self, node: Option<usize>) {
        let deadline = tokio::time::Instant::now() + STARTUP_TIMEOUT;
        loop {
            let ready = match node {
                Some(replica_id) => {
                    log_contains(&self.nodes[replica_id].log_path, "replica mesh complete")
                }
                None => self
                    .nodes
                    .iter()
                    .all(|process| log_contains(&process.log_path, "replica mesh complete")),
            };
            if ready {
                return;
            }
            assert!(
                tokio::time::Instant::now() < deadline,
                "Iggy cluster did not form its replica mesh before the deadline"
            );
            tokio::time::sleep(Duration::from_millis(100)).await;
        }
    }
}

impl Drop for TestIggyCluster {
    fn drop(&mut self) {
        self.proxy_task.abort();
        for node in &mut self.nodes {
            let _ = node.child.kill();
            let _ = node.child.wait();
        }
    }
}

async fn spawn_stable_proxy(
    proxy_port: u16,
    target: Arc<AtomicU16>,
) -> tokio::task::JoinHandle<()> {
    let listener = tokio::net::TcpListener::bind(("127.0.0.1", proxy_port))
        .await
        .expect("bind stable cluster endpoint");
    tokio::spawn(async move {
        while let Ok((mut incoming, _)) = listener.accept().await {
            let target_port = target.load(Ordering::Acquire);
            tokio::spawn(async move {
                if let Ok(mut upstream) =
                    tokio::net::TcpStream::connect(("127.0.0.1", target_port)).await
                {
                    let _ = tokio::io::copy_bidirectional(&mut incoming, &mut upstream).await;
                }
            });
        }
    })
}

#[allow(dead_code)]
pub struct TestIggy {
    binary: PathBuf,
    child: Mutex<Child>,
    data_dir: TempDir,
    tcp_port: u16,
}

#[allow(dead_code)]
impl TestIggy {
    pub async fn start() -> Self {
        Self::start_inner(free_host_port()).await
    }

    pub async fn start_pinned() -> Self {
        Self::start_inner(free_host_port()).await
    }

    async fn start_inner(tcp_port: u16) -> Self {
        let binary = resolve_server_binary();
        let data_dir = tempfile::tempdir().expect("create Iggy test data directory");
        let child = spawn_server(&binary, data_dir.path(), tcp_port);
        let server = Self {
            binary,
            child: Mutex::new(child),
            data_dir,
            tcp_port,
        };
        server.wait_until_ready().await;
        server
    }

    pub async fn restart(&self) {
        {
            let mut child = self.child.lock().expect("lock Iggy test process");
            child.kill().expect("kill Iggy test process");
            child.wait().expect("reap Iggy test process");
            *child = spawn_server(&self.binary, self.data_dir.path(), self.tcp_port);
        }
        self.wait_until_ready().await;
    }

    pub async fn client(&self) -> Result<IggyClient, IggyError> {
        let client = IggyClientBuilder::new()
            .with_tcp()
            .with_server_address(format!("127.0.0.1:{}", self.tcp_port))
            .build()?;
        client.connect().await?;
        client
            .login_user(DEFAULT_ROOT_USERNAME, DEFAULT_ROOT_PASSWORD)
            .await?;
        Ok(client)
    }

    pub async fn laser(&self, stream: impl Into<String>) -> Result<Laser, IggyError> {
        Ok(Laser::from_client(self.client().await?).with_default_stream(stream))
    }

    pub async fn laser_reconnecting(&self, stream: impl Into<String>) -> Result<Laser, IggyError> {
        let client = IggyClientBuilder::from_connection_string(&format!(
            "iggy+tcp://{DEFAULT_ROOT_USERNAME}:{DEFAULT_ROOT_PASSWORD}@127.0.0.1:{}",
            self.tcp_port
        ))?
        .build()?;
        client.connect().await?;
        Ok(Laser::from_client(client).with_default_stream(stream))
    }

    async fn wait_until_ready(&self) {
        let deadline = tokio::time::Instant::now() + STARTUP_TIMEOUT;
        loop {
            if self
                .child
                .lock()
                .expect("lock Iggy test process")
                .try_wait()
                .expect("inspect Iggy test process")
                .is_some()
            {
                let log = std::fs::read_to_string(self.log_path())
                    .unwrap_or_else(|error| format!("failed to read server log: {error}"));
                panic!(
                    "Iggy test server exited during startup, log {}:\n{log}",
                    self.log_path().display(),
                );
            }
            let ready = tokio::time::timeout(Duration::from_secs(2), self.probe_writes())
                .await
                .map(|result| result.is_ok())
                .unwrap_or(false);
            if ready {
                return;
            }
            if tokio::time::Instant::now() >= deadline {
                let log = std::fs::read_to_string(self.log_path())
                    .unwrap_or_else(|error| format!("failed to read server log: {error}"));
                panic!(
                    "Iggy did not accept VSR writes within 30s, log {}:\n{log}",
                    self.log_path().display(),
                );
            }
            tokio::time::sleep(Duration::from_millis(150)).await;
        }
    }

    async fn probe_writes(&self) -> Result<(), IggyError> {
        let client = self.client().await?;
        let stream = format!("ready_{}", std::process::id());
        let topic = "probe";
        if client
            .get_stream(&Identifier::named(&stream)?)
            .await?
            .is_none()
        {
            client.create_stream(&stream).await?;
        }
        let stream_id = Identifier::named(&stream)?;
        let topic_id = Identifier::named(topic)?;
        if client.get_topic(&stream_id, &topic_id).await?.is_none() {
            let options = TopicCreateOptions {
                partitions_count: Some(1),
                compression_algorithm: Some(CompressionAlgorithm::default()),
                message_expiry: Some(IggyExpiry::NeverExpire),
                max_topic_size: Some(MaxTopicSize::ServerDefault),
                ..TopicCreateOptions::default()
            };
            client.create_topic(&stream_id, topic, &options).await?;
        }
        let producer = client.producer(&stream, topic)?.build();
        producer.init().await?;
        let message = IggyMessage::builder()
            .payload(Bytes::from_static(b"ready"))
            .build()?;
        producer.send(vec![message]).await.map(|_| ())
    }

    fn log_path(&self) -> PathBuf {
        self.data_dir.path().join("test-server.log")
    }
}

impl Drop for TestIggy {
    fn drop(&mut self) {
        let child = self.child.get_mut().expect("lock Iggy test process");
        let _ = child.kill();
        let _ = child.wait();
    }
}

fn spawn_server(binary: &Path, data_dir: &Path, tcp_port: u16) -> Child {
    let log_path = data_dir.join("test-server.log");
    let stdout = File::create(&log_path).expect("create Iggy test server log");
    let stderr = stdout.try_clone().expect("clone Iggy test server log");
    Command::new(binary)
        .env("IGGY_SYSTEM_PATH", data_dir)
        .env("IGGY_TCP_ADDRESS", format!("127.0.0.1:{tcp_port}"))
        .env("IGGY_HTTP_ENABLED", "false")
        .env("IGGY_QUIC_ENABLED", "false")
        .env("IGGY_WEBSOCKET_ENABLED", "false")
        .env("IGGY_ROOT_USERNAME", DEFAULT_ROOT_USERNAME)
        .env("IGGY_ROOT_PASSWORD", DEFAULT_ROOT_PASSWORD)
        .env("IGGY_SHARD_RUNTIME_CAPACITY", "256")
        .env("IGGY_SYSTEM_SHARDING_CPU_ALLOCATION", "all")
        .env("IGGY_SYSTEM_SHARDING_RECONCILE_PERIODIC_INTERVAL", "200 ms")
        .stdout(Stdio::from(stdout))
        .stderr(Stdio::from(stderr))
        .spawn()
        .expect("spawn the VSR Iggy test server")
}

fn spawn_cluster_node(
    binary: &Path,
    data_dir: &Path,
    log_path: &Path,
    replica_id: usize,
    tcp_ports: &[u16; CLUSTER_SIZE],
    replica_ports: &[u16; CLUSTER_SIZE],
) -> Child {
    let stdout = File::create(log_path).expect("create cluster node log");
    let stderr = stdout.try_clone().expect("clone cluster node log");
    let mut command = Command::new(binary);
    command
        .arg("--replica-id")
        .arg(replica_id.to_string())
        .env("IGGY_SYSTEM_PATH", data_dir)
        .env("IGGY_CLUSTER_ENABLED", "true")
        .env("IGGY_CLUSTER_NAME", "laser-sdk-rolling-restart")
        .env("IGGY_MESSAGE_BUS_RECONNECT_PERIOD", "100ms")
        .env("IGGY_HTTP_ENABLED", "false")
        .env("IGGY_QUIC_ENABLED", "false")
        .env("IGGY_WEBSOCKET_ENABLED", "false")
        .env("IGGY_ROOT_USERNAME", DEFAULT_ROOT_USERNAME)
        .env("IGGY_ROOT_PASSWORD", DEFAULT_ROOT_PASSWORD)
        .env("IGGY_SHARD_RUNTIME_CAPACITY", "256")
        .env("IGGY_SYSTEM_SHARDING_CPU_ALLOCATION", "0..1")
        .env("IGGY_SYSTEM_SHARDING_RECONCILE_PERIODIC_INTERVAL", "200 ms")
        .stdout(Stdio::from(stdout))
        .stderr(Stdio::from(stderr));
    for node in 0..CLUSTER_SIZE {
        command
            .env(
                format!("IGGY_CLUSTER_NODES_{node}_NAME"),
                format!("node-{node}"),
            )
            .env(format!("IGGY_CLUSTER_NODES_{node}_IP"), "127.0.0.1")
            .env(
                format!("IGGY_CLUSTER_NODES_{node}_REPLICA_ID"),
                node.to_string(),
            )
            .env(
                format!("IGGY_CLUSTER_NODES_{node}_PORTS_TCP"),
                tcp_ports[node].to_string(),
            )
            .env(
                format!("IGGY_CLUSTER_NODES_{node}_PORTS_TCP_REPLICA"),
                replica_ports[node].to_string(),
            );
    }
    command.spawn().expect("spawn Iggy cluster node")
}

fn log_contains(path: &Path, marker: &str) -> bool {
    std::fs::read_to_string(path).is_ok_and(|content| content.contains(marker))
}

fn graceful_stop(child: &mut Child) {
    // SAFETY: `child.id()` is the live process spawned and owned by this harness.
    unsafe {
        libc::kill(child.id() as libc::pid_t, libc::SIGTERM);
    }
    let deadline = std::time::Instant::now() + Duration::from_secs(5);
    while std::time::Instant::now() < deadline {
        if child.try_wait().expect("inspect cluster node").is_some() {
            return;
        }
        std::thread::sleep(Duration::from_millis(20));
    }
    child.kill().expect("kill cluster node after grace period");
    child.wait().expect("reap cluster node");
}

fn resolve_server_binary() -> PathBuf {
    if let Some(path) = std::env::var_os(IGGY_SERVER_ENV) {
        let path = PathBuf::from(path);
        assert!(
            path.is_file(),
            "{IGGY_SERVER_ENV} is not a file: {}",
            path.display()
        );
        return path;
    }

    let resolver = repository_root()
        .join("scripts")
        .join("resolve-test-iggy-server.sh");
    let output = Command::new(&resolver)
        .output()
        .expect("run the Iggy test server resolver");
    assert!(
        output.status.success(),
        "Iggy test server resolver failed: {}",
        String::from_utf8_lossy(&output.stderr)
    );
    PathBuf::from(
        String::from_utf8(output.stdout)
            .expect("resolver path is UTF-8")
            .trim(),
    )
}

fn repository_root() -> PathBuf {
    let manifest = Path::new(env!("CARGO_MANIFEST_DIR"));
    if manifest.ends_with("sdk") {
        return manifest
            .parent()
            .expect("sdk has a repository parent")
            .to_owned();
    }
    manifest
        .parent()
        .and_then(Path::parent)
        .expect("bdd/rust has a repository parent")
        .to_owned()
}

fn free_host_port() -> u16 {
    std::net::TcpListener::bind("127.0.0.1:0")
        .expect("bind an ephemeral port")
        .local_addr()
        .expect("read the bound port")
        .port()
}

fn free_host_ports(count: usize) -> Vec<u16> {
    let listeners = (0..count)
        .map(|_| std::net::TcpListener::bind("127.0.0.1:0").expect("reserve host port"))
        .collect::<Vec<_>>();
    listeners
        .iter()
        .map(|listener| listener.local_addr().expect("reserved address").port())
        .collect()
}
